segmentation_labeling_app/rois/rois.py

```diff
@@ -92,102 +195,86 @@ def roi_from_query(cls, roi_id: int) -> "ROI":
+        absolute_threshold: float
+            weighted entries above this value=1, below=0
+            over-ridden if cumulative_threshold is set
+        cumulative_threshold: float
+            if specified, set absolute threshold determined by fraction of
+            total weight above this value
+        nbins: int
+            if cumulative specified, determines
+            for interpolating
+        dilation_kernel_size: int
+            passed as size to cv2.getStructuringElement()
+        inner_outline: bool
+            whether to retain only outline points interior to the mask
+
+        Returns
+        -------
+        mask: numpy.ndarray
+            2D dense representation of the mask outline
+
         """
-        if optional_array is None:
-            optional_array = self.generate_binary_mask_from_threshold(threshold)
-        contours, _ = cv2.findContours(optional_array,
+        binary = binary_mask_from_threshold(
+            self._sparse_coo,
+            absolute_threshold=absolute_threshold,
+            cumulative_threshold=cumulative_threshold,
+            nbins=nbins)
+
+        contours, _ = cv2.findContours(binary,
                                        cv2.RETR_LIST,
                                        cv2.CHAIN_APPROX_NONE)
-        if len(contours) > 0:
-            return np.unique(np.squeeze(np.concatenate(contours), axis=1), axis=0)
-        else:
-            logging.warning(msg=f"no edges detected for ROI: {self.roi_id} with"
-                            f" ophys_exp_id: {self.experiment_id}")
-            return np.array([])
-
-    def get_edge_mask(self, threshold: float,
-                      optional_array: np.array = None) -> np.array:
-        """
-        Returns a mask of edge points where edges of an roi are represented
-        as 1 and non edges are 0. Computes edges using binary erosion and
-        exclusive or operation. Generates edges from a thresholded binary mask.
-        Args:
-            threshold: a float to threshold all values against to create binary
-                       mask
-            optional_array: an optional array to overload the function and
-            create edge points from different mask
 
-        Returns:
-            mask_matrix: a binary mask of the edges given a threshold
+        xy = np.concatenate(contours).squeeze()
 
-        """
-        edge_coordinates = self.get_edge_points(threshold=threshold,
-                                                optional_array=optional_array)
-        mask_matrix = np.zeros(shape=self.image_shape)
-        if len(edge_coordinates) > 0:
-            for edge_coordinate in edge_coordinates:
-                mask_matrix[edge_coordinate[1]][edge_coordinate[0]] = 1
-        return mask_matrix
+        mask = np.zeros_like(binary)
+        mask[xy[:, 1], xy[:, 0]] = 1.0
+
+        kernel = np.ones((dilation_kernel_size, dilation_kernel_size))
+
+        mask = cv2.dilate(mask, kernel)
```

Comment:
Do we want to specify iterations on the dilation? If dilation iterations are specified the bitwise and will not work for subtracting away non outline mask pixels.

Reply:
I think the default of 1 is fine. This is mainly for visualization.
I think the AND is working

Reply:
For clarification, the AND won't work with iterations > 1, as we expand away from the original mask it will fail where we have expanded the borders. But that's not relevant if we're capping the iterations at 1.

Reply:
still don't know what you mean. This is what I would expect:
```
In [220]: binary                                                                   
Out[220]: 
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)

In [221]: contours, _ = cv2.findContours(binary, cv2.RETR_LIST, cv2.CHAIN_APPROX_NO
     ...: NE)                                                                      

In [222]: xy = np.concatenate(contours).squeeze()                                  

In [223]: mask = np.zeros_like(binary)                                             

In [224]: mask[xy[:, 1], xy[:, 0]] = 1.0                                           

In [225]: mask                                                                     
Out[225]: 
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 1, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 1, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 1, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 1, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 1, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)

In [226]: mask = cv2.dilate(mask, kernel)                                          

In [227]: mask                                                                     
Out[227]: 
array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 0, 0, 0, 0, 1, 1, 1],
       [1, 1, 1, 0, 0, 0, 0, 1, 1, 1],
       [1, 1, 1, 0, 0, 0, 0, 1, 1, 1],
       [1, 1, 1, 0, 0, 0, 0, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]], dtype=uint8)

In [228]: mask = mask & binary                                                     

In [229]: mask                                                                     
Out[229]: 
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 1, 0, 0, 0, 0, 1, 1, 0],
       [0, 1, 1, 0, 0, 0, 0, 1, 1, 0],
       [0, 1, 1, 0, 0, 0, 0, 1, 1, 0],
       [0, 1, 1, 0, 0, 0, 0, 1, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)
```

Reply:
Isaak I think the source of confusion here could be that Dan is dilating the mask, not the outline.

Reply:
I expect the same thing to happen as what you posted with the AND. Unless I'm mistaken the borders are supposed to expand outward with multiple iterations not inwards towards the center, so the AND subtracts away those border pixels. What I would expect from the dilation_iterations==2 would be the mask below, which the AND function does not provide us.
```
      [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 0, 0, 0, 0, 0, 0, 1, 1],
       [1, 1, 0, 0, 0, 0, 0, 0, 1, 1],
       [1, 1, 0, 0, 0, 0, 0, 0, 1, 1],
       [1, 1, 0, 0, 0, 0, 0, 0, 1, 1],
       [1, 1, 0, 0, 0, 0, 0, 0, 1, 1],
       [1, 1, 0, 0, 0, 0, 0, 0, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]
```
I think this is the expansion direction because otherwise the border would begin to overlap the ROI and it would be harder for the labeler to see. But correct me if I'm wrong.

Reply:
Ah, I see the misunderstanding. `dilate` doesn't know anything about inside or outside. It just dilates:
https://docs.opencv.org/4.2.0/d4/d86/group__imgproc__filter.html#ga4ff0f3318642c4f469d0e11f242f3b6c

Reply:
Right, and we would be expanding our borders inward with each iteration using the AND not outwards. This was the motivation for the iterative egde detection and dilating in my previous ROIs class, I didn't know a way to do that process more efficiently.

Reply:
I'm still not sure what the complaint is here. Again, this is just for visualization, but, what Kat wanted was thicker borders that don't extend outwards past the original ROI boundary. She wanted the "inner outline"

Reply:
@kschelonka I was under the impression the borders were not supposed to overlap the ROI so they should expand outward from the 'inner outline'. In my mind if they expand inward the outline will overlap the ROI which will lead to a bad labeling experience. Could you clarify for me which way the ROI edges are to expand? Thanks!

Reply:
oh, if I misunderstood that, we can just change to `mask = mask & (not binary)`

Reply:
I think we still want the original border and if there's a hole that will add the entire hole to the mask, if I'm not mistaken. But I could be wrong, I'll go and test that and see if it comes across okay.

Reply:
Dan is correct about what I meant by "inner outline". My thought originally was that we want the outline to be bounded to the actual edge of the cell, and not extend past it. If we generate a few examples of each way I'm open to considering an outline that would extend beyond the cell. Since the outlines are not going to be very thick it probably won't matter a great deal.

Reply:
The user will have the option to view a transparent mask overlay or an outline.

Reply:
That's my bad for the confusion, still confusion over which way we are expanding lol. I think we should expand outward as we don't want to intersect the actual ROI, I think it'll make it hard to view. If you both are on the same page about borders expanding inward I'm okay with that.